kubelet bootstrap: start hostNetwork pods before we have PodCIDR

justinsb opened the pull request:
Network readiness was checked in the pod admission phase, but pods that
fail admission are not retried.  Move the check to the pod start phase.

Issue #35409 
Issue #35521

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35526)

<!-- Reviewable:end -->

justinsb committed:
Create hostNetwork pods even if network plugin not ready

We do now admit pods (unlike the first attempt), but now we will stop
non-hostnetwork pods from starting if the network is not ready.

Issue #35409

justinsb committed:
Update bringup: don't pass in dummy pod-cidr

We no longer pass in a "dummy" pod-cidr (10.123.45.0/29), and rely on
reconcile-cidr=true instead (which is the default).

justinsb commented:
This should fix the problem that needed a reversion of #33347 (static pods that fail admission are not rescheduled).  Should I apply this on top of #33347 and squash for review?

cc @yujuhong @dchen1107 

Also not clear whether I should move the disk space check out of the admission check.  I'd lean towards doing it separately.

yujuhong commented:
> static pods that fail admission because network-not-ready are not rescheduled

All regular pods got rejected whenever kubelet restarts was a equally serious problem :-)

> Also not clear whether I should move the disk space check out of the admission check. I'd lean towards doing it separately.

I think we actually want to reject pods if disk is full, so not sure if we want to change that, and definitely not in this PR.
As for the not retrying the static pods, we will fix that separately (reusing #35521 for this purpose). 


> Should I apply this on top of #33347 and squash for review?

Sounds good. Thanks!

justinsb commented:
> All regular pods got rejected whenever kubelet restarts was a equally serious problem :-)

Ooops... sorry!

I rebased on top of the prior (now reverted) PR.  Looks much simpler now!

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1311,7 +1311,14 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
-	if errOuter := canRunPod(pod); errOuter != nil || pod.DeletionTimestamp != nil || apiPodStatus.Phase == api.PodFailed {
+	errOuter := canRunPod(pod)
+	if errOuter == nil {
+		rs := kl.runtimeState.networkErrors()
```
/cc @bprashanth I don't think this works with the CRI integration, in which case, we'll just sync all the pods and fail?

justinsb replied:
Not entirely sure what you're saying  :-) .  We now start synchronizing before the network is ready, so that we can see the hostNetwork pods.  But we can't actually start pods that don't use hostNetwork, because the network plug is not ready.

justinsb replied:
Oh I think I understand...

Perhaps if we renamed to `networkPluginErrors`, or `podNetworkErrors`?  The idea is that if the network plugin has a problem, we put it into this separate list.  An error that prevents hostNetwork pods would presumably not be a network plugin error, and would go into the general errors list (which prevents the pod sync loop from even starting)

justinsb commented on pkg/kubelet/kubelet.go:
```diff
@@ -1311,7 +1311,14 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	if errOuter == nil {
+		rs := kl.runtimeState.networkErrors()
+		if len(rs) != 0 && !podUsesHostNetwork(pod) {
+			errOuter = fmt.Errorf("Network is not ready: %v", rs)
```
Fixed the error nit - thanks.

Can you explain when we do & don't want to kill the pod?  Is it the case that we kill if the failure is permanent, but not if the failure is transient?

yujuhong replied:
AFAIK, we only kill pods if the pod is getting deleted (`DeletionTimestamp` is set), evicted, or if kubelet is not capable (in terms of security context) to run the pod at all. Although I am not sure how kubelet could've started the pod for the latter case to begin with...

yujuhong replied:
Most of the pod that kubelet determined not to run would be rejected at admission, and not entering syncpod at all.

justinsb commented:
Rebased.

@yujuhong are you happy with this, if we we renamed to `networkPluginErrors`, or `podNetworkErrors`?  Any preference?

zmerlynn commented:
I'd like to get this PR in soon, since it's blocking the `kops-updown` test from passing (and has for a while).

yujuhong commented:
@justinsb I am fine with the PR. Please squash the commits.

justinsb commented:
@yujuhong  squashed, but also addressed https://github.com/kubernetes/kubernetes/pull/35526#discussion_r85049149 which I worry that I had not addressed - per your suggestion,  I just return error instead of returning the pod.  Thanks!

dchen1107 commented:
LGTM, but can I have a unittest to capture the scenario: Network is not ready, all pods are blocked except the one using hostNetwork?

justinsb committed:
kubelet: testing for hostNetwork with network errors

justinsb commented:
@dchen1107 now with added unit-test :-)

zkmoney commented:
@justinsb if this passes all tests, when do you expect this fix to be available in kubelet?

edevil commented:
@zkmoney I think this is dependant on a Kubernetes member applying the LGTM label.

FWIW, LGTM to me. :)

justinsb commented:
Assuming it gets LGTM-ed before freeze, it should be in 1.5.  I don't know if we would backport it .. we should probably add a flag if we did.

zkmoney commented:
Ah got it, thank you. I am eagerly awaiting the ability to remove a dummy,
broken --pod-cidr value from my bootstrap kubelet sysconfig and still have
the master be able to start the api server pods (I run all of the
components via manifests).

If there is any good solution to this that you know of in the meantime, I
would be grateful. Thanks.


On Fri, Nov 4, 2016 at 2:06 PM Justin Santa Barbara <
notifications@github.com> wrote:

> Assuming it gets LGTM-ed before freeze, it should be in 1.5. I don't know
> if we would backport it .. we should probably add a flag if we did.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/35526#issuecomment-258505821>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATM25fLEHpbObgeYxIVn7mmPg8avgR35ks5q63QjgaJpZM4KgI3e>
> .
>

edevil commented:
Yes, I'm in the same boat of the dummy pod-cidr. And this is a problem for daemon-set pods because they are started on the master nodes. I can try to cherry-pick this for 1.4 if someone sets the appropriate label.

edevil commented:
On another note, the submit queue bot seems to have missed the lgtm label...

zkmoney commented:
@edevil FWIW for daemonsets where they don't actually need to be running on the masters I've used the `spec.template.spec.nodeSelector` to target nodes with the appropriate labels and avoid masters, but I know this isn't a viable solution for all or long term.

edevil commented:
@zkmoney This is for the logging component (Fluentd) so I do want it to run on every host. :)

dchen1107 commented:
LGTM

edevil commented:
Can this be backported to 1.4?

zkmoney commented:
👍 For that @edevil
